### ISSUE TITLE
jsf server application

### ISSUE THREAD

**Issue by ducin** [json-schema-faker/json-schema-faker#145]
Dear All,

I gathered authors of three libraries here:
* [`raml-server`](https://github.com/farolfo/raml-server/): @farolfo, @Jastrzebowski
* [`raml-mocker`](https://github.com/repocho/raml-mocker/): @repocho, @biggestT
* [`json-schema-faker`](https://github.com/json-schema-faker/json-schema-faker): @pateketrueke and me, @ducin
and @galk-in who works with RAML and might also be interested in this topic.

I can see we all want to achieve the same - and that is **RESTful API stub server automation**. Define RAML file and run a stub server out of the box (with no JS code). I'm writing to you all to suggest that we all join forces and integrate our tools. Our tools are responsible for:
* `json-schema-faker` - extends JSON Schema draft 4, provides built-in support for `faker.js` and `chance.js` generator functions. All features of both faker and chance are available in jsf; see our [demo](json-schema-faker.js.org)
* `raml-mocker` - interprets RAML spec and provides a way to use `faker.js` through hand-made formats; can produce list of request/response objects out from a RAML spec
* `raml-server` - uses `raml-mocker` to deploy a real REST API, also uses json-server to simulate persistence (data state, if I can call it that)

I can also see that:
* `raml-server` relies on `raml-mocker`; `json-schema-faker` is not used by them
* `raml-mocker` and `json-schema-faker` both use faker.js, but `json-schema-faker` can do a lot more  in that matter; but `raml-mocker`also interprets RAML spec and jsf doesn't do that (and will not)
* although `raml-server` uses `raml-mocker`, extending JSON definitions with faker.js seems to be cumbersome at this point; no chance.js support as well
* `json-schema-faker` is significantly more popular than the other 2 libraries, but is not integrated with them. Yet ;)
* finally, some of the issues solved by `raml-mocker` and `json-schema-faker` are exactly the same ones, so it doesn't make sense to carry on duplicating our efforts.

What I would like to consider is that **we integrate all our three libraries into one damn awesome tool stack that could serve data**! *Together we are stronger* and I believe that we will surely achieve more if we join forces! We were thinking about serving jsf-made data from a real API (actually the work is already started: [json-schema-server](https://github.com/json-schema-faker/json-schema-server) but not much work has been put there **yet**).

Are you interested in joining forces to create an awesome tool stack together?

**Comment by pateketrueke** [comment]
Also: https://github.com/gextech/raml-mockup

**Comment by ducin** [comment]
Forgot about that one. Seems like it does some of the stuff that `raml-server` does... Anyway, I'd suggest to reduce the repetition, integrate our tools and conquer swagger! :beers:

**Comment by Jastrzebowski** [comment]
@ducin great idea! As for now I use both **raml-mockup** as API generator and **json-schema-faker** in components tests, would be great to have one tool for both cases :)

**Comment by repocho** [comment]
@ducin Yes ! I' willing to collaborate putting all of this in one common component. I don't have much time right now but I will do my best.

**Comment by farolfo** [comment]
@ducin I'm on board! Let's unite forces to make something bigger ;)

**Comment by galk-in** [comment]
Hello. Let's share wish-list about the mock-project.
I'm using the mock for end-2-end testing. So repeatability is important: 
1) Persistant in memory.
2) Possibility to restart the mock with same initial data.
3) [jsonpatch](http://jsonpatch.com)

**Comment by ducin** [comment]
Gentlemen,

I'm very happy that you *all* responded positive to my initiative :) Nobody refused, everybody's in, perfect :grinning:
I'm sorry for the long reply, but I had my laptop motherboard burnt and later I was ill, just recovering but I've been thinking intensively about all this. Below are my thoughts. I'm very excited and I hope we'll start something amazing together :)

Please keep in mind all the below are my ideas and I don't want to force them. If you disagree, I'm 100% opened for discussion, let's bring something better up together :) there are some pieces of the puzzle, existing already, but they're not as powerful as they could be, when integrated.

# the main goal

The main goal I'm thinking of is to provide an **amazing tool stack that will read a RAML contract and run a stub REST API server _out of the box_**. This is something I'm sure community will :cupid: I'm sure about that :heart_eyes: 

## present

This is how I see [`raml-server`](https://github.com/farolfo/raml-server/) & [`raml-mocker`](https://github.com/repocho/raml-mocker/) today:
![image](https://cloud.githubusercontent.com/assets/375027/15126051/7e6d3c54-1630-11e6-8826-ad0811001492.png)

* [`raml-server`](https://github.com/farolfo/raml-server/) reads a RAML contract and runs a stub REST API
* [`json-server`](https://github.com/typicode/json-server) provides REST API engine that handles GET, POST, PUT, DELETE requests
* [`raml-mocker`](https://github.com/repocho/raml-mocker/) interprets RAML, interprets some of json schema, uses faker.js through manual format functions

This is current [`json-schema-faker`](https://github.com/json-schema-faker/json-schema-faker):
![image](https://cloud.githubusercontent.com/assets/375027/15126118/c3fa2f16-1630-11e6-8a71-8dee008749e3.png)

* supports most of json schema draft 4
* uses 3 generator libraries (faker, chance, randexp)
* provides some custom features such as default behaviors, custom formats, extending 3rd party libs; open for new features

## future

The sketch below is what I'm thinking of:
![image](https://cloud.githubusercontent.com/assets/375027/15126084/a57bebce-1630-11e6-8595-98dcb378051a.png)

The architectural change it includes is:
* reshaping `raml-mocker`:
  * drops interpretation of json schema - let `json-schema-faker` do that (its support is significantly bigger)
  * drops support for custom formats which are easier under `json-schema-faker`
  * uses `json-schema-faker` as a dependency.

The reasons behind above ideas are that:
* `json-schema-faker` and `raml-mocker` duplicate responsibilities and it makes completely no sense in the long run. I'd opt to let json-schema-faker deal with json schema and random libraries since it has bigger support, not because it's the library I write (please trust me here! :) ). `raml-mocker`, however, remains responsible for consuming RAML contracts in order to support REST API / HTTP request-response stuff
* `raml-mocker`, after having learned to use `json-schema-faker`, can support more json schema out of the box.
* If we ever need to improve json schema support, we do it in one place :) instead of two

## community

I believe that it's always better to let other people come in and help us. And for that reason I'd opt to create a github organisation that would hold this concept. If I got back in time and npm registry was empty, I'd choose `raml-server` as the perfect name for the tool. So I'd opt to have `raml-server` as the name of the organisation. And all of us who wish to contribute would get access there.

I wish that `raml2html` (documentation) and `raml-server/raml-server` (REST API stub server) will be the obvious choices for RAML community :)

# sum-up

key objectives:
- **main goal**: provide a tool (entire tool stack beneath) that turns a RAML contract into a working HTTP REST API based on stubs
- remove feature duplication between tools
- integrate all tools into one stack
- create github org for the stack to make it consistent
- provide docs so that others can help us easier

----

Please share your feedback guys! What do you agree with? What would you do in a different way?

When we all agree, we could split big tasks into smaller pieces to get the things going!

----

regarding the wish-list @galk-in gave - it's a good idea. Let's decide whether these make more sense on `json-schema-faker` or `raml-mocker`. Do you have a suggestion?

**Comment by galk-in** [comment]
RAML describe only interface of the REST API, but we need some logic for making a valuable mock.
In abao we have hooks, where you can very easy set some logic for requests. I think we should make the same. 

So my idea is hooks for raml-server with logic for express function.

Now i'm using something like this in one our project. In future i will delivery it to open-source.

**Comment by repocho** [comment]
> What do you agree with?

All !! I do agree all !!

I think we need some kind of organization in the community to share the works to achieve that proposal. 

For the moment I've drop the dereferencing functionality from `raml-mocker` because now it can be done in `raml-js-parser`

I think the next step is to drop all the json generation and give `json-schema-faker` this responsibility. But we need to define a kind of roadmap or tasks to be done. I do not have much time so I need to have a backlog of small tasks to be able to organize myself and the project.

So what are the next steps ?

**Comment by repocho** [comment]
Hi guys !!

Any news on that ?? @ducin please, could you lead that and set the new steps ?

We should be prepared also for the new RAML 1.0 schema.,

**Comment by pateketrueke** [comment]
Guys,

- I've recently updated `raml-mockup` for RAML 1.0 support
- I've recently stabilized `json-schema-faker` and now we can go beyond
